libtextsecure/message_receiver.js

```diff
@@ -1545,24 +1546,42 @@ MessageReceiver.prototype.extend({
       window.log.info('Error getting conversation: ', envelope.source);
     }
   },
+  async handleSessionRequestMessage(envelope, content) {
+    const shouldProcessSessionRequest = await session.Protocols.SessionProtocol.shouldProcessSessionRequest(
+      envelope.source,
+      envelope.timestamp
+    );
+
+    if (shouldProcessSessionRequest) {
+      if (content.preKeyBundleMessage) {
+        await this.savePreKeyBundleMessage(
+          envelope.source,
+          content.preKeyBundleMessage
+        );
+      }
+      await session.Protocols.SessionProtocol.onSessionRequestProcessed(
+        envelope.source
+      );
+      window.log.info('sending session established to', envelope.source);
+      // We don't need to await the call below because we just want to send it off
+      window.libloki.api.sendSessionEstablishedMessage(envelope.source);
+    }
+  },
   async innerHandleContentMessage(envelope, plaintext) {
     const content = textsecure.protobuf.Content.decode(plaintext);
+    const { SESSION_REQUEST } = textsecure.protobuf.Envelope.Type;
 
-    if (content.preKeyBundleMessage) {
-      await this.savePreKeyBundleMessage(
-        envelope.source,
-        content.preKeyBundleMessage
+    if (envelope.type === SESSION_REQUEST) {
+      await this.handleSessionRequestMessage(envelope, content);
```

Comment:
Do you want `handleFriendRequestAcceptIfNeeded` below to also be called or should we return after the session's been processed? Session requests messages have no other payload, right?

Reply:
We should still continue processing the rest of the message. For example in the case of an `END_SESSION` message, we want to let the code below handle it.

Reply:
Do we make `SESSION_REQUEST` and `END_SESSION` compatible? To me it sounds like the two should be handled on the same level, or even collapsed into the same command.

Reply:
yes, they both get funnelled into `handleSessionRequestMessage` since both have the envelope type `SESSION_REQUEST`. However like i said in the case of `END_SESSION` (or another message we might want in the future) we don't have the `null` field set on the message, so it can pass through to the other handlers.